File: elixir/test/app_identity_test.exs
defmodule AppIdentityTest do
  use AppIdentity.Case, async: true
  doctest AppIdentity

  for version <- AppIdentity.Versions.supported() do
    test "generate v#{version} proof", context do
      app_config = context[unquote(version)]

      assert {:ok, proof} = AppIdentity.generate_proof(app_config)

      assert_generate_proof_telemetry_span(app_config, proof: proof)

      case unquote(version) do
        1 -> assert [_, _, _] = decode_to_parts(proof)
        _ -> assert [_, _, _, _] = decode_to_parts(proof)
      end
    end

    test "generate valid v#{version} proof", context do
      app_config = context[unquote(version)]
      app = AppIdentity.App.new!(app_config)

      assert {:ok, proof} = AppIdentity.generate_proof(app_config)

      assert_generate_proof_telemetry_span(app_config, proof: proof)

      assert {:ok, verified(app)} == AppIdentity.verify_proof(proof, app_config)

      assert_verify_proof_telemetry_span(app, proof)
    end
  end

  test "verify fails if not base64", %{v1: v1} do
    assert :error == AppIdentity.verify_proof("not base64", v1)
    assert_verify_proof_telemetry_span(v1, "not base64", error: "cannot decode proof string")
  end

  test "verify fail on insufficent parts", %{v1: v1} do
    proof = Base.url_encode64("a:b")

    assert :error == AppIdentity.verify_proof(proof, v1)

    assert_verify_proof_telemetry_span(v1, proof,
      error: "proof must have 3 parts (version 1) or 4 parts (any version)"
    )
  end

  test "verify fail on bad v1 nonce", %{v1: v1, v1_app: v1_app} do
    padlock = build_padlock(v1_app, nonce: "n:once")
    proof = build_proof(v1_app, padlock, nonce: "n:once")

    assert :error == AppIdentity.verify_proof(proof, v1)

    assert_verify_proof_telemetry_span(v1, proof,
      error: "version cannot be converted to a positive integer"
    )
  end

  test "verify fail on bad v2 nonce format", %{v1: v1, v1_app: v1_app} do
    padlock = build_padlock(v1_app)
    proof = build_proof(v1_app, padlock, version: 2)

    assert :error == AppIdentity.verify_proof(proof, v1)
    assert_verify_proof_telemetry_span(v1, proof, error: "nonce does not look like a timestamp")
  end

  test "verify fail on v2 nonce out of fuzz", %{v1: v1, v1_app: v1_app} do
    nonce = timestamp_nonce(-11, :minutes)
    padlock = build_padlock(v1_app, nonce: nonce)
    proof = build_proof(v1_app, padlock, version: 2, nonce: nonce)

    assert :error == AppIdentity.verify_proof(proof, v1)
    assert_verify_proof_telemetry_span(v1, proof, error: "nonce is invalid")
  end

  test "verify fail on v1 nonce for v2 app", %{v2: v2, v2_app: v2_app} do
    padlock = build_padlock(v2_app, version: 1)
    proof = build_proof(v2_app, padlock, version: 1)

    assert :error == AppIdentity.verify_proof(proof, v2)
    assert_verify_proof_telemetry_span(v2, proof, error: "proof and app version mismatch")
  end

  for padlock_case <- [:upper, :lower] do
    test "verify success v1 (padlock case #{padlock_case})", %{v1: v1, v1_app: v1_app} do
      padlock = build_padlock(v1_app, case: unquote(padlock_case))

      proof = build_proof(v1_app, padlock)

      assert {:ok, verified(v1_app)} == AppIdentity.verify_proof(proof, v1)
      assert_verify_proof_telemetry_span(v1, proof)
    end

    test "verify success v2 default fuzz (padlock case #{padlock_case})", %{
      v2: v2,
      v2_app: v2_app
    } do
      nonce = timestamp_nonce(-6, :minutes)

      padlock = build_padlock(v2_app, nonce: nonce, case: unquote(padlock_case))

      proof = build_proof(v2_app, padlock, version: 2, nonce: nonce)

      assert {:ok, verified(v2_app)} == AppIdentity.verify_proof(proof, v2)
      assert_verify_proof_telemetry_span(v2, proof)
    end

    test "verify success v2 custom fuzz (padlock case #{padlock_case})" do
      v2 = v2(fuzz: 300)
      {:ok, v2_app} = AppIdentity.App.new(v2)

      nonce = timestamp_nonce(-2, :minutes)
      padlock = build_padlock(v2_app, nonce: nonce, case: unquote(padlock_case))
      proof = build_proof(v2_app, padlock, version: 2, nonce: nonce)

      assert verified(v2_app) == AppIdentity.verify_proof!(proof, v2)

      assert_verify_proof_telemetry_span(v2, proof)
    end
  end

  test "verify fail on different app ids", %{v1: v1, v1_app: v1_app} do
    padlock = build_padlock(v1_app, id: "00000000-0000-0000-0000-000000000000")
    proof = build_proof(v1_app, padlock, id: "00000000-0000-0000-0000-000000000000")

    assert :error == AppIdentity.verify_proof(proof, v1)
    assert_verify_proof_telemetry_span(v1, proof, error: "proof and app do not match")
  end

  test "verify fail on bad padlock", %{v1: v1, v1_app: v1_app} do
    padlock = build_padlock(v1_app, nonce: "foo")
    proof = build_proof(v1_app, padlock)

    assert {:ok, nil} == AppIdentity.verify_proof(proof, v1)

    assert_verify_proof_telemetry_span(v1, proof, app: :none)
  end

  test "verify fail on non-hex padlock", %{v1: v1, v1_app: v1_app} do
    padlock =
      v1_app
      |> build_padlock()
      |> String.replace(~r/[A-F]/i, "z")

    proof = build_proof(v1_app, padlock)

    assert :error == AppIdentity.verify_proof(proof, v1)

    assert_verify_proof_telemetry_span(v1, proof, error: "padlock must be a hex string value")
  end
end

File: ts/test/app_identity.test.ts
import base64url from 'base64-url'
import { expect, test } from 'vitest'

import { App } from '../src/app'
import { generateProofWithDiagnostic, verifyProofWithDiagnostic } from '../src/internal'
import * as Support from '../support'

import '../support/matchers'

test('generate v1 proof', () => {
  const proof = generateProofWithDiagnostic(Support.v1Input())
  expect(Support.decodeToParts(proof).length).toEqual(3)
})

test('generate valid v1 proof', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)

  const proof = generateProofWithDiagnostic(v1)
  expect(verifyProofWithDiagnostic(proof, v1)).toBeVerified(app)
})

test('generate v2 proof', () => {
  const proof = generateProofWithDiagnostic(Support.v2Input())
  expect(Support.decodeToParts(proof).length).toEqual(4)
})

test('generate valid v2 proof', () => {
  const v2 = Support.v2Input()
  const app = new App(v2)

  const proof = generateProofWithDiagnostic(v2)
  expect(verifyProofWithDiagnostic(proof, v2)).toBeVerified(app)
})

test('generate valid v3 proof', () => {
  const v3 = Support.v3Input()
  const app = new App(v3)

  const proof = generateProofWithDiagnostic(v3)
  expect(verifyProofWithDiagnostic(proof, v3)).toBeVerified(app)
})

test('generate valid v4 proof', () => {
  const v4 = Support.v4Input()
  const app = new App(v4)

  const proof = generateProofWithDiagnostic(v4)
  expect(verifyProofWithDiagnostic(proof, v4)).toBeVerified(app)
})

test('verify fails if not base64', () => {
  expect(() => verifyProofWithDiagnostic('not base64', Support.v1Input())).toThrowError(
    'proof must have 3 parts (version 1) or 4 parts (any version)'
  )
})

test('verify fail on insufficient parts', () => {
  expect(() =>
    verifyProofWithDiagnostic(base64url.encode('a:b'), Support.v1Input())
  ).toThrowError('proof must have 3 parts (version 1) or 4 parts (any version)')
})

test('verify fail on bad v1 nonce', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)
  const padlock = Support.buildPadlock(app, { nonce: 'n:once' })
  const proof = Support.buildProof(app, padlock, { nonce: 'n:once' })

  expect(() => verifyProofWithDiagnostic(proof, v1)).toThrowError(
    'version cannot be converted to an integer'
  )
})

test('verify fail on bad v2 nonce format', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)
  const padlock = Support.buildPadlock(app)
  const proof = Support.buildProof(app, padlock, { version: 2 })

  expect(() => verifyProofWithDiagnostic(proof, v1)).toThrowError(
    'nonce does not look like a timestamp'
  )
})

test('verify fail on v2 nonce out of fuzz', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)
  const nonce = Support.timestampNonce(-11, 'minutes')
  const padlock = Support.buildPadlock(app, { nonce })
  const proof = Support.buildProof(app, padlock, { version: 2, nonce })

  expect(() => verifyProofWithDiagnostic(proof, v1)).toThrowError('nonce is invalid')
})

test('verify fail on v1 nonce for v2 app', () => {
  const v2 = Support.v2Input()
  const app = new App(v2)
  const padlock = Support.buildPadlock(app, { version: 1 })
  const proof = Support.buildProof(app, padlock, { version: 1 })

  expect(() => verifyProofWithDiagnostic(proof, v2)).toThrowError(
    'proof and app version mismatch'
  )
})

test('verify success v1', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)
  const padlock = Support.buildPadlock(app)
  const proof = Support.buildProof(app, padlock)

  expect(verifyProofWithDiagnostic(proof, v1)).toBeVerified(app)
})

test('verify success v2 default fuzz', () => {
  const v2 = Support.v2Input()
  const app = new App(v2)
  const nonce = Support.timestampNonce(-6, 'minutes')
  const padlock = Support.buildPadlock(app, { nonce })
  const proof = Support.buildProof(app, padlock, { nonce })
  expect(verifyProofWithDiagnostic(proof, v2)).toBeVerified(app)
})

test('verify success v2 custom fuzz', () => {
  const v2 = Support.v2Input(300)
  const app = new App(v2)
  const nonce = Support.timestampNonce(-2, 'minutes')
  const padlock = Support.buildPadlock(app, { nonce })
  const proof = Support.buildProof(app, padlock, { nonce })

  expect(verifyProofWithDiagnostic(proof, v2)).toBeVerified(app)
})

test('verify fail on different app ids', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)
  const id = '00000000-0000-0000-0000-000000000000'
  const padlock = Support.buildPadlock(app, { id })
  const proof = Support.buildProof(app, padlock, { id })

  expect(() => verifyProofWithDiagnostic(proof, v1)).toThrowError(
    'proof and app do not match'
  )
})

test('verify fail on bad padlock', () => {
  const v1 = Support.v1Input()
  const app = new App(v1)
  const padlock = Support.buildPadlock(app, { nonce: 'foo' })
  const proof = Support.buildProof(app, padlock)

  expect(verifyProofWithDiagnostic(proof, v1)).toBeNull()
})
